Updated sample to use new TLS Api

lbosquett opened the pull request:
# Updated sample to use new TLS Api

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [ ] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->

Updated the `Certificate.Optional.Sample` to use the new API and with host names instead of ip addresses.

## Description

I've changed the sample code to use the new method `public static ListenOptions UseHttps(this ListenOptions listenOptions, TlsHandshakeCallbackOptions callbackOptions)` to configure when the client certificate is required to make the handshake. Also changed the README to reflect the current state of the sample

## Details

+ Load the server certificate using the store, as this is required to create the instance of `SslServerAuthenticationOptions`, I got the `System.NotSupportedException: The server mode SSL must use a certificate with the associated private key` without the `ServerCertificate` property specified
+ in `README.md` and `Program.cs`: Changed ip address to host names `example.com` and `cert.example.com`, also changed the `mydomain` to `example.com`, see [IANA-managed Reserved Domain](https://www.iana.org/domains/reserved)
+ Made the implementation inside `OnConnection` callback, tried to mantain the same behavior but I saw that `ClientCertificateMode` is not avaliable anymore, so I used the `ClientCertificateRequired` together with `RemoteCertificateValidationCallback` to block the connection.

## Tests

I followed the docs to generate a [self siged certificate through powershell](https://docs.microsoft.com/en-us/dotnet/core/additional-tools/self-signed-certificates-guide) and generated a certificate for `example.com` and `cert.example.com` (This can be a inline doc in Program.cs to instruct user about the sub domain)

Changed the `hosts` file inside `C:\Windows\System32\drivers\etc` with the content:

```txt
127.0.0.1 example.com
127.0.0.1 cert.example.com
```

Used the powershell to store the certificate in `cert:\LocalMachine\Root`, restarted chrome and tried to access the https://example.com:5001 (you'll see the screenshots with the 5005 port because I tested with it, but I changed the final code to 5001 later).

![image](https://user-images.githubusercontent.com/46539053/145940613-9270cc49-ad5b-4fae-8b89-f9d2a3aa6f50.png)

now, let's try `https://cert.example.com`:

![image](https://user-images.githubusercontent.com/46539053/145940660-95deb5bf-eeef-40da-bb4d-e6f1d4f227a2.png)

✔ I choosed cancel and got the `ERR_BAD_SSL_CLIENT_AUTH_CERT` as expected:

![image](https://user-images.githubusercontent.com/46539053/145940698-601fff47-d0c4-4bc8-956d-24a34e30e7f2.png)

✔ tried again and have choosen the certificate, successfully worked:

![image](https://user-images.githubusercontent.com/46539053/145940882-faed9919-8eff-4e8c-82c7-f16f4bb7ad21.png)

## Notes

❌ the `/auth` seems not working and keeps showing the `Too many redirects` message in navigator for me, I think this isn't related to my changes since the old code had the same behavior.

![image](https://user-images.githubusercontent.com/46539053/145941207-5e00bf03-a89e-46be-a187-590fbb0243b6.png)


Fixes [#37753](https://github.com/dotnet/aspnetcore/issues/37753)

lbosquett committed:
Updated sample to use new TLS Api

Tratcher commented on src/Security/Authentication/Certificate/samples/Certificate.Optional.Sample/Program.cs:
```diff
@@ -10,8 +12,8 @@ namespace Certificate.Optional.Sample;
 
 public class Program
 {
-    public const string HostWithoutCert = "127.0.0.1";
-    public const string HostWithCert = "127.0.0.2";
+    public const string HostWithoutCert = "example.com";
+    public const string HostWithCert = "cert.example.com";
```
These values didn't need to change, and it actually breaks running the sample unless someone adds HOST file entries for them. Please revert, and keep that note in the readme.

lbosquett commented on src/Security/Authentication/Certificate/samples/Certificate.Optional.Sample/Program.cs:
```diff
@@ -22,23 +24,38 @@ public static void Main(string[] args)
+                    options.ListenAnyIP(5001, listenOptions =>
                     {
-                        listenOptions.UseHttps(httpsOptions =>
+                        listenOptions.UseHttps(new TlsHandshakeCallbackOptions()
```
Important: see the old code and the issue https://github.com/dotnet/runtime/issues/31097
with the new API looks unnecessary to have the two calls to Listen because now every connection will make a call to `OnConnection ` instead of initializing a "state" for every connection; but I'm not 100% sure about this

Tratcher replied:
You're correct, only one Listen is needed here.

lbosquett committed:
Updated sample to use new TLS Api

Reverted README.md to original
Reverted changes from domains back to 127.0.0.1 and 127.0.0.2

Tratcher commented:
Thanks @lbosquett
